sdk/authentication.py

```diff
@@ -0,0 +1,154 @@
+
+        Args:
+            token_response: A valid token response dictionary.
+
+        Returns:
+            The time the token will expire, or None if it can't be calculated
+        """
+        expires_in = token_response.get("expires_in", None)
+        if expires_in:
+            return datetime.now(tz=timezone.utc) + timedelta(seconds=float(expires_in))
+
+        try:
+            # We assume the token is valid, but might not be in an expected format
+            token = token_response.get("access_token")
+
+            if isinstance(token, str) and "." in token:
+                token_timestamp = decode(
+                    token,
+                    algorithms=["HS256"],
+                    options={"verify_signature": False},
+                ).get("exp")
+                if token_timestamp:
+                    return datetime.fromtimestamp(token_timestamp, tz=timezone.utc)
+
+        except DecodeError:
+            # The token is not in a format that could be parsed as a JWT
+            pass
+        return None
+
+
+class Auth0TokenProvider(TokenProvider):
+    def __init__(self, username: str, password: str):
+        super().__init__(username, password)
+
+        #   This is public, so we can put it in the code here.
+        self.public_client_id = "PeZvo7Atx7IVv3iel59asJSb4Ig7vuSB"
+        self.audience = Endpoints.account
+        self.realm = "pcs-users"
+        self.domain = Endpoints.auth0_domain
+
+        # Makes a call in order to check the credentials at creation
+        self._query_token()
+
+    def _query_token(self) -> dict[str, Any]:
+        token = GetToken(self.domain)
+
+        # No client secret required for this Application since
+        # "Token Endpoint Authentication Method" set to None
+        validated_token: dict[str, Any] = token.login(
+            client_id=self.public_client_id,
+            client_secret="",
+            username=self.username,
+            password=self.password,
+            audience=self.audience,
+            scope="openid profile email",
+            realm=self.realm,
+            grant_type="http://auth0.com/oauth/grant-type/password-realm",
+        )
+        return validated_token
+
+
+class FakeAuth0GoodAuthentication(TokenProvider):
+    def _query_token(self) -> dict[str, Any]:
+        return {
+            "access_token": "some_token",
+            "id_token": "id_token",
+            "scope": "openid profile email",
+            "expires_in": 86400,
+            "token_type": "Bearer",
+        }
+
+
+class FakeAuth0BadAuthentication(TokenProvider):
+    def __init__(self, username: str, password: str):
+        super().__init__(username, password)
+
+        #   Makes a call in order to check the credentials at creation
+        #   And raise the Error
+        self._query_token()
+
+    def _query_token(self) -> dict[str, Any]:
+        raise Auth0Error(
+            status_code=403, error_code=403, message="Wrong email or password"
+        )
```

Comment:
I assume this is used for testing purposes ? Can we move this to the tests folder ?

Reply:
To me, it's good practice to have those implementations here so that, if we make a modification to the class above, they need to be reflected on those here. :)
But if that sounds like a problem to you, I'll move them

Reply:
I tried moving them, I had a few issues and I simply didn't want to adress them since we agreed with Matthieu in private that this solution was acceptable as well and both implementations were OK
so, we can say it's ''done''